scipy/spatial/distance.py

```diff
@@ -1395,6 +1451,10 @@ def dfun(u, v):
 
     elif isinstance(metric, string_types):
         mstr = metric.lower()
+        if w is not None and not mstr.startswith("test_"):
+            if mstr in ['seuclidean', 'se', 's', 'mahalanobis']:
+                raise ValueError("metric %s incompatible with weights" % mstr)
+            mstr = "test_%s" % mstr
```

Comment:
This is required since the C implementations do not implement weights. This PR adds weights to all the distance functions. I'm not a C expert, so I think it's good for me to stop here, but have these weighted version in python. If someone wants to come in and make the C version weight-aware, then we don't need this.